src/SpanContainer.cpp

```diff
@@ -1,47 +1,66 @@
-#include "SpanContainer.h"
+#include <forward_list>
 #include "Core/src/Assert.h"
+#include "SpanContainer.h"
 
 using namespace std;
 using namespace core;
 
 namespace cppkin
 {
+    using SpansList = std::forward_list<Span*>;
+
 #if defined(WIN32)
-    static __declspec(thread) Span* m_span = nullptr; //Not unique ptr due to limitation on static initialization for TLS object.
+    static __declspec(thread) SpansList* m_spans = nullptr; //Not unique ptr due to limitation on static initialization for TLS object.
 #else
 
 #  include <features.h>
 #  if __GNUC_PREREQ(4,8)
-    static thread_local Span* m_span = nullptr;
+    static thread_local SpansList* m_spans = nullptr;
+    static thread_local std::unique_ptr<Span::SpanHeader> m_rootHeader = nullptr;
 #else
-    static __thread Span* m_span = nullptr;
+    static __thread SpansList* m_spans = nullptr;
+    static __thread std::unique_ptr<Span::SpanHeader> m_rootHeader = nullptr;
 #endif
 #endif
 
+    static SpansList* GetSpans() {
+        if (!m_spans){
+            m_spans = new std::forward_list<Span*>();
+        }
+        return m_spans;
+    }
+
     SpanContainer& SpanContainer::Instance(){
         static SpanContainer instance;
         return instance;
     }
 
-    SpanContainer::~SpanContainer(){
-        delete m_span;
-    }
+    SpanContainer::SpanContainer(){}
+    SpanContainer::~SpanContainer(){}
 
-    Span& SpanContainer::GetSpan() const{
-        return *m_span;
+    const Span* SpanContainer::TopSpan() const{
+        if (GetSpans()->empty()){
+            return nullptr;
+        }
+        return GetSpans()->front();
     }
 
-    SpanContainer::SpanContainer(){}
+    const Span::SpanHeader* SpanContainer::GetRootHeader() const{
+        return m_rootHeader.get();
+    }
 
-    void SpanContainer::SetSpan(Span* span) {
-        swap(m_span, span);
-        delete span;
+    void SpanContainer::PushSpan(Span* span) {
+        GetSpans()->push_front(span);
+        if(span->IsRootSpan())
+            m_rootHeader.reset( new Span::SpanHeader(span->GetHeader()));
     }
 
-    unique_ptr<Span> SpanContainer::ReleaseSpan(){
+   Span* SpanContainer::PopSpan(){
```

Comment:
Should return unique_ptr to reflect transition of ownership. if empty, throw an exception.